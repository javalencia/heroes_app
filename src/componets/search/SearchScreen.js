import React, { useMemo } from 'react';
import queryString from 'query-string';
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../selectors/getHeroesByName';


export const SearchScreen = ({ history }) => {

    const location = useLocation();

    const { q = '' } = queryString.parse( location.search );

    const [  formValues, handleInputChange ] = useForm({ 
        searchText: q 
    });
    
    const { searchText } = formValues;
    
    //const heroesFiltered = getHeroesByName( searchText );

    const heroesFiltered = useMemo(() =>  getHeroesByName( q ), [ q ])

    const handleSubmit = ( e ) => {
        e.preventDefault();

        history.push(`?q=${ searchText }`)

    }

    return (
        <>
            <h3>Search Screen</h3>
            <hr />

            <div className="row">

                <div className="col-5">
                <h4> Search Form</h4>
                    <form onSubmit={ handleSubmit }>
                        <input 
                            type="text"
                            placeholder="Find your here"
                            className="form-control"
                            name="searchText"
                            value={ searchText }
                            onChange={ handleInputChange }
                            autoComplete="off"
                        />

                        <button
                            type="submit"
                            className="btn m1 btn-block btn-outline-primary"
                        >
                            Search..
                        </button>

                    </form>
                </div>

                <div className="col-7">

                    <h4> Results </h4>
                    <hr />

                    {
                        ( q === '')
                            &&
                            <div className="alert alert-info">
                                Search Superhero
                            </div>

                    }

                    {
                        ( q !== ''  && heroesFiltered.length === 0 )
                            &&
                            <div className="alert alert-danger">
                                There is not hero with { q }
                            </div>

                    }

                    {
                        heroesFiltered.map( hero => (
                            <HeroCard 
                                key={ hero.id }
                                { ...hero }
                            />
                        ))
                    }


                </div>

            </div>
        </>
    )
}
