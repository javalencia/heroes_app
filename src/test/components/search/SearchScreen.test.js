import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../componets/search/SearchScreen';

describe('Pruebas en el <SearchScreen />', () => {

    test('Debe de mostrarse correctamente con valores por defecto', () => {

        const wrapper = mount(
            <MemoryRouter  initialEntries={['/search']} >
               <Route 
                 path="/search" 
                 component={ SearchScreen } 
                />
            </MemoryRouter>
        );


        expect( wrapper ).toMatchSnapshot();

        expect( wrapper.find('.alert-info').text().trim() ).toEqual('Search Superhero');

    });


    test('Debe de mostar a batman y el input con el valor del Query String', () => {
       
        const wrapper = mount(
            <MemoryRouter  initialEntries={['/search?q=batman']} >
               <Route 
                 path="/search" 
                 component={ SearchScreen } 
                />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toEqual( 'batman' )
        expect( wrapper ).toMatchSnapshot();

    });

    test('debe de mostrar un error si no se encuentra el hero', () => {
        

        const q = 'batman123456'

        const wrapper = mount(
            <MemoryRouter  initialEntries={[`/search?q=${ q }`]} >
               <Route 
                 path="/search" 
                 component={ SearchScreen } 
                />
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').text().trim() ).toEqual(`There is not hero with ${ q }`);
    });

    test('debe de llamar el push del history', () => {
       
        const history = { 
            push: jest.fn()
            
        }

        const wrapper = mount(
            <MemoryRouter  initialEntries={[`/search?q=batman123`]} >
               <Route 
                 path="/search" 
                 component={() =>  <SearchScreen history={ history } /> } 
                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target:{
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect( history.push ).toHaveBeenCalledWith(`?q=batman`)


    });
    
    
    
    
  
});
