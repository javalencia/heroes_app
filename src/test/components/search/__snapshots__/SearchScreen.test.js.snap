// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen /> Debe de mostar a batman y el input con el valor del Query String 1`] = `
Array [
  <h3>
    Search Screen
  </h3>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your here"
          type="text"
          value="batman"
        />
        <button
          className="btn m1 btn-block btn-outline-primary"
          type="submit"
        >
          Search..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                 
                Batman
                 
              </h5>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                   
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en el <SearchScreen /> Debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h3>
    Search Screen
  </h3>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your here"
          type="text"
          value=""
        />
        <button
          className="btn m1 btn-block btn-outline-primary"
          type="submit"
        >
          Search..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search Superhero
      </div>
    </div>
  </div>,
]
`;
